src/charm.py

```diff
@@ -73,37 +94,112 @@ def _on_jenkins_pebble_ready(self, event: PebbleReadyEvent) -> None:
         """Configure and start Jenkins server.
 
         Args:
-            event: Event fired when pebble is ready.
+            event: The event fired when pebble is ready.
         """
         container = event.workload
         if not container or not container.can_connect():
             event.defer()
             return
 
-        self.unit.status = MaintenanceStatus("Configuring Jenkins.")
+        self.unit.status = MaintenanceStatus("Installing Jenkins.")
         # First Jenkins server start installs Jenkins server.
         container.add_layer(
-            "jenkins", self._get_pebble_layer(calculate_env(admin_configured=False)), combine=True
+            "jenkins",
+            self._get_pebble_layer(jenkins.calculate_env(admin_configured=False)),
+            combine=True,
         )
         container.replan()
         try:
-            wait_jenkins_ready()
-            unlock_jenkins(container)
+            jenkins.wait_ready()
+            self.unit.status = MaintenanceStatus("Configuring Jenkins.")
+            jenkins.bootstrap(
+                container,
+                self.state.jnlp_port,
```

Comment:
Perhaps these could be grouped in some form so that this doesn't need to pass the specific items and rather make an object available that has all the required data available? That reduces the amount of coupling between charm.py and jenkins.py

Reply:
A lot of the previous states/config values have been removed, with `jnlp_port` being the only value remaining. Perhaps when there are more options available, it can be refactored to `BootstrapConfig`.